lib/polyfill/pip.js

```diff
@@ -0,0 +1,230 @@
+        return polyfillPictureInPictureElement;
+      },
+
+      set(value) {
+        if (value === polyfillPictureInPictureElement) return;
+
+        if (polyfillPictureInPictureElement) {
+          polyfillPictureInPictureElement.removeEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument_,
+          );
+        }
+
+        polyfillPictureInPictureElement =
+          /** @type {HTMLMediaElement} */ (value);
+        if (polyfillPictureInPictureElement) {
+          polyfillPictureInPictureElement.addEventListener(
+            'webkitpresentationmodechanged',
+            shaka.polyfill.PiP.updatePictureInPictureElementInDocument_,
+          );
+        }
+      },
+    });
+
+    /**
+     * polyfill document.pictureInPictureEnabled
+     */
+    document.pictureInPictureEnabled = true;
+
+    /**
+     * polyfill HTMLMediaElement.requestPictureInPicture
+     */
+    proto.requestPictureInPicture = shaka.polyfill.PiP.requestPictureInPicture_;
+
+
+    /**
+     * polyfill document.exitPictureInPicture
+     */
+    document.exitPictureInPicture = shaka.polyfill.PiP.exitPictureInPicture_;
+
+    /**
+     * polyfill enterpictureinpicture and leavepictureinpicture events
+     */
+
+    const oldAddEventListener = proto.addEventListener;
+    /**
+     * @this {HTMLMediaElement}
+     */
+    proto.addEventListener = function(type, listener, options) {
+      if (type === 'enterpictureinpicture') {
+        const proxyEnterPiPEvent =
+          shaka.polyfill.PiP.getProxyEnterPiPEvent_(listener);
+        this.addEventListener('webkitpresentationmodechanged',
+          proxyEnterPiPEvent);
+      } else if (type === 'leavepictureinpicture') {
+        const proxyLeavePiPEvent =
+          shaka.polyfill.PiP.getProxyLeavePiPEvent_(listener);
+        this.addEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldAddEventListener.apply(this, [type, listener, options]);
+      }
+    };
+
+    const oldRemoveEventListener = proto.removeEventListener;
+    /**
+     * @this {HTMLMediaElement}
+     */
+    proto.removeEventListener = function(type, listener, options) {
+      if (type === 'enterpictureinpicture') {
+        const proxyEnterPiPEvent =
+          shaka.polyfill.PiP.getProxyEnterPiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyEnterPiPEvent);
+      } else if (type === 'leavepictureinpicture') {
+        const proxyLeavePiPEvent =
+          shaka.polyfill.PiP.getProxyLeavePiPEvent_(listener);
+        this.removeEventListener('webkitpresentationmodechanged',
+          proxyLeavePiPEvent);
+      } else {
+        // fallback for all the other events
+        oldRemoveEventListener.apply(this, [type, listener, options]);
+      }
+    };
+  }
+};
+
+/**
+ * @this {HTMLMediaElement}
```

Comment:
I don't think this annotation is correct.  I think this event listener will not be called in the context of a media element.  Is that right?

Reply:
Yes. I changed it.